### DISCUSSION TITLE
Support of industry standard QL/GPM - openCypher - GQL

### DISCUSSION THREAD

**Discussion by annainfo** [terminusdb/terminusdb#1657] [Ideas]
TerminusDB currently doesn't seem to support querying via industry standard openCypher nor the ISO/IEC DIS GMP.

Background:
https://opencypher.org/
"openCypher is the most widely adopted, fully-specified, and open query language for property graph databases."
and
"It provides an intuitive way to work with property graphs today and is the best on-ramp to [the graph query language (GQL) standard being developed by ISO](https://www.gqlstandards.org/)."

https://aws.amazon.com/blogs/database/announcing-the-general-availability-of-opencypher-support-for-amazon-neptune/
2022-04-02 
"Announcing the General Availability of openCypher support for Amazon Neptune"

https://www.tigergraph.com/press-article/tigergraph-announces-commitment-to-support-opencypher-in-gsql/
2022-10-19
"In addition, TigerGraph’s openCypher support aligns with support for industry standard GQL"

https://github.com/opencypher/cypher-for-gremlin
"Cypher for Gremlin is a toolkit for users of [Apache TinkerPop™](https://tinkerpop.apache.org/) that allows querying Gremlin databases with [Cypher](https://neo4j.com/docs/developer-manual/current/cypher/), the industry's most widely used [property graph](https://github.com/opencypher/openCypher/blob/master/docs/property-graph-model.adoc) query language defined and maintained by the [openCypher](http://www.opencypher.org/) project."

https://www.gqlstandards.org/existing-languages
"The lineage of existing graph query languages and the SQL/PGQ and GQL standards work. Note the overlap between SQL/PGQ (Property Graph Queries in SQL) and GQL (Graph Query Language). In the ISO/IEC standards, the specifications that are used by both parts will be defined in one standard and referenced by the other standard. Source: Petra Selmer"

https://www.iso.org/standard/79473.html
ISO/IEC DIS 9075-16
Information technology — Database languages SQL — Part 16: Property Graph Queries (SQL/PGQ)
stage 40.60

https://www.iso.org/standard/76120.html
ISO/IEC CD 39075.2
Information Technology — Database Languages — GQL
stage 30.60

Since the GPM is shared between the two ISO/IEC standards under development, the first standard published, SQL/PGQ has the higher stage value, will contain the definition of the GPM. The GPM is close to openCypher.

If TerminusDB would allow querying via openCypher 
1) a switch from other DBs that currently allow that is easier
2) there would be less of a lock-in to adopters of TerminusDB

https://www.iso.org/stages-and-resources-for-standards-development.html
Enquiry stage (40)
"If the DIS is approved and no technical changes are introduced in the draft, the project goes straight to publication. However, if technical changes are introduced, FDIS stage is mandatory."

It is possible, that the GPM as described in the current ISO/IEC DIS 9075-16 is already the version that will become the ISO/IEC GPM.

TerminusDB could buy the DIS and compare it with openCypher to determine how close the two are and report back here.

**Comment by annainfo** [comment]
https://www.gqlstandards.org/ July, 2022 – GQL Status Update
"SQL/PGQ is interesting to people focused on GQL because the Graph Pattern Matching (GPM) language in SQL/PGQ is essentially identical to GPM in GQL. Since the technical work on SQL/PGQ is now stable, GPM in GQL is now stable."

TerminusDB could start working on support of GPM. And possibly be ready when the standard is published
https://ldbcouncil.org/event/fifteenth-tuc-meeting/attachments/keith-hare-property-graph-standards-process-and-timing.pdf

"Expected Dates
• DIS Ballot completes – December 2022
• Published standard – June 2023"

**Comment by luke-feeney** [comment]
We basically agree with all these points and the work to support GraphQL is very much designed to facilitate those switching from other DBs and to reduce lock-in. We will continue to try to add more support over time, including for openCypher and/or GQL.  As mentioned in the other comment, we think we need to build a combinator library in rust. This will facilitate query in GraphQL, utility queries so we can move some of the startup out of prolog, and then eventually could be used to implement GQL.

**Comment by hpvd** [comment]
+1 on this!

in addition:
just a link to what is possible with terminusDB's WOQL (Web Object Query Language)
and how it's done:
https://terminusdb.com/blog/the-power-of-web-object-query-language/

For the reason: "why" I found this:
"curious why a new query language WOQL was designed and implemented instead of just using...?"
answer: https://news.ycombinator.com/item?id=22871961

full context:
https://news.ycombinator.com/item?id=22867767

**Comment by hpvd** [comment]
Maybe this includes some interesting ideas when looking into supporting opencypher:
Bridging graph data models: RDF, RDF-star, and property graphs as directed acyclic graphs
https://arxiv.org/abs/2304.13097

**Comment by hpvd** [comment]
since we are talking about opencypher which is typical used in the fied of property graphs,
it looks interesting to me (probably this is an old topic for you) that  one could lower the possible advantages of property graphs vs RDFs
in the fields of
- dealing with properties on edges
- being good at graph traversals

by looking into the possibilities of RDF-star for adding metadata and SPARQL versions 1.1+ for path support,
while keeping all the goodness of RDF.

**Comment by hpvd** [comment]
supporting open cypher brings also an other advantage to the table:
it would be far easier to get used to terminusdb in the field of AI, e.g.
- when following examples for building things like RAG or
- using terminusdb directly in langchain

which would give potential new users a good chance to discover the unique functionalities of terminusdb...

**Reply by annainfo** [reply]
GQL is FDIS, expected to become published as a standard 2024 Q2.

If TerminusDB still has interest in supporting GQL - as claimed in the roadmap since more than a year - then it could adopt it directly.

**Reply by hpvd** [reply]
thanks for update on this!

(since I have to google for it:  FDIS is the current Approval Stage - Final Draft International Standard (FDIS))

**Reply by hpvd** [reply]
Basic GQL (edit: not GQL -> GraphQL) seems to be available since TerminusDB 10.1.8
https://github.com/terminusdb/terminusdb/releases/tag/v10.1.8
and from then there was an ongoing refinement, see release notes
https://github.com/terminusdb/terminusdb/releases?q=Graphql&expanded=true
and blog posts:
https://terminusdb.com/?s=GraphQL

**Reply by annainfo** [reply]
The links refer to GraphQL, not GQL.

**Reply by hpvd** [reply]
ok, thanks. Since I only dealt with GraphQL before, I thought GQL is referring to this...

**Comment by annainfo** [comment]
It's over, no acitivity, the Cypher part isn't of interest anymore. ISO GQL is in FDIS stage. 

Idea focused on ISO GQL and ISO SQL/PGQ which share the GPM at https://github.com/orgs/terminusdb/discussions/2088

**Reply by alexander-mart** [reply]
_Copy from [duplicated topic](https://github.com/orgs/terminusdb/discussions/2088):_

## Implement GQL:2024 (FDIS stage, ETA 2024-04/2024-05) or at least the SELECT/MATCH part of SQL:2023 Part 16 SQL/PGQ which shares the GPM.

1. 2019-09-10 ISO GQL - stage 10.99 New project approved https://www.iso.org/standard/76120.html
2. [pre 2022-03-30, couldn't find a date] "A Bit later" at https://github.com/orgs/terminusdb/projects/2
3. 2023-01-17 (a year and two weeks ago)
3.1) "Support of industry standard QL/GPM - openCypher - GQL" - including info about stages of GQL, SQL/PGQ and the shared GPM https://github.com/orgs/terminusdb/discussions/1657
3.2) "GPM is stable - any news regarding GQL support?" including answer by a maintainer: no capacity, and "We think we need to build a combinator library in rust." https://github.com/orgs/terminusdb/discussions/1662
4. 2023-06-01 SQL:2023 published including "ISO/IEC 9075-16:2023, Information technology, Database languages SQL, Part 16: Property Graph Queries (SQL/PGQ)" https://www.iso.org/standard/79473.html
5. 2023-06-20 message by maintainer "We don't have active plans to add SPARQL support, but it would not be a particularly hard project to do so." https://github.com/orgs/terminusdb/discussions/830
6. 2023-10-16 "WOQL - rewrite in order" - no mention of GQL https://github.com/orgs/terminusdb/discussions/1996
7. 2024-01-16 "The publication process takes six to eight weeks so ISO/IEC 39075:2024 Information technology — Database languages — GQL should officially be published in late April or early May of 2024!" https://www.gqlstandards.org/
8. 2024-01-26 GQL:2024 at stage 50:20 https://www.iso.org/standard/76120.html
